internal/driverutil/description.go

```diff
@@ -0,0 +1,899 @@
+// Copyright (C) MongoDB, Inc. 2024-present.
+//
+// Licensed under the Apache License, Version 2.0 (the "License"); you may
+// not use this file except in compliance with the License. You may obtain
+// a copy of the License at http://www.apache.org/licenses/LICENSE-2.0
+
+package driverutil
+
+import (
+	"errors"
+	"fmt"
+	"math"
+	"time"
+
+	"go.mongodb.org/mongo-driver/bson"
+	"go.mongodb.org/mongo-driver/event"
+	"go.mongodb.org/mongo-driver/internal/bsonutil"
+	"go.mongodb.org/mongo-driver/internal/handshake"
+	"go.mongodb.org/mongo-driver/internal/ptrutil"
+	"go.mongodb.org/mongo-driver/mongo/address"
+	"go.mongodb.org/mongo-driver/mongo/readpref"
+	"go.mongodb.org/mongo-driver/tag"
+	"go.mongodb.org/mongo-driver/x/mongo/driver/description"
+)
+
+// CompositeServerSelector combines multiple selectors into a single selector by
+// applying them in order to the candidates list.
+//
+// For example, if the initial candidates list is [s0, s1, s2, s3] and two
+// selectors are provided where the first matches s0 and s1 and the second
+// matches s1 and s2, the following would occur during server selection:
+//
+// 1. firstSelector([s0, s1, s2, s3]) -> [s0, s1]
+// 2. secondSelector([s0, s1]) -> [s1]
+//
+// The final list of candidates returned by the composite selector would be
+// [s1].
+type CompositeServerSelector struct {
```

Comment:
Optional: Does it make sense to move all the server selector logic to its own internal package? It would at least make naming friendlier.

For example,
```go
driverutil.CompositeServerSelector
```
could become
```go
serverselector.Composite
```

Note: Generally more packages increases the chances for import cycles, but the risk of changing the structure of internal packages is low because it's always reversible.